osprepare/distro.go

```diff
@@ -0,0 +1,268 @@
+	var sources []*aptSourcesFile
+
+	// Closure is only relevant to this function
+	addAptSources := func(path string) {
+		if r := readAptSourcesFile(path); r != nil {
+			sources = append(sources, r)
+		}
+	}
+
+	if pathExists(aptSourcesList) {
+		addAptSources(aptSourcesList)
+	}
+
+	// Glob the *.list files now
+	if pathExists(aptSourcesListD) {
+		tpath := fmt.Sprintf("%s/*.list", aptSourcesListD)
+		if files, err := filepath.Glob(tpath); err == nil {
+			for _, file := range files {
+				addAptSources(file)
+			}
+		}
+	}
+
+	return sources
+}
+
+// newAptSource constructs a new apt source from
+// the given deb style line
+func newAptSource(debLine string) *aptSource {
+	fields := strings.Fields(debLine)
+	var asource aptSource
+
+	if len(fields) < 3 {
+		return nil
+	}
+
+	dtype := fields[0]
+	if dtype != "deb" && dtype != "deb-src" {
+		return nil
+	}
+
+	asource.DebType = dtype
+	asource.Origin = fields[1]
+	asource.Distribution = fields[2]
+	if len(fields) > 3 {
+		asource.Components = fields[3:]
+	}
+
+	return &asource
+}
+
+// isUbuntuDockerRepoEnabled iterates through the sources
+// to find out if the docker repo is enabled or not.
+func isUbuntuDockerRepoEnabled() bool {
+	sources := loadAptSources()
+	if sources == nil {
+		return false
+	}
+	for _, sourceFile := range sources {
+		for _, source := range sourceFile.Sources {
+			if strings.Contains(source.Origin, "apt.dockerproject.org") {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// pathExists is a helper function which handles the
+// error and simply return true or false if the given
+// path exists
+func pathExists(path string) bool {
+	if _, err := os.Stat(path); err != nil {
+		return false
+	}
+	return true
+}
+
+type distro interface {
+	// InstallPackages should implement the installation
+	// of packages using distro specific methods for
+	// the given target list of items to install
+	InstallPackages(packages []string) bool
+
+	// getID should return a string specifying
+	// the distribution ID (e.g: "clearlinux")
+	getID() string
+}
+
+// getDistro will return a distro based on what
+// is read from GetOsRelease
+func getDistro() distro {
+	osRelease := getOSRelease()
+
+	if osRelease == nil {
+		return nil
+	}
+
+	if strings.HasPrefix(osRelease.ID, "clear-linux") {
+		return &clearLinuxDistro{}
+	} else if strings.Contains(osRelease.ID, "ubuntu") {
+		// Store the Ubuntu codename, i.e. "xenial'
+		return &ubuntuDistro{CodeName: osRelease.GetValue("UBUNTU_CODENAME")}
+	} else if strings.Contains(osRelease.ID, "fedora") {
+		return &fedoraDistro{}
+	}
+	return nil
+}
+
+// os-release clear-linux*
+type clearLinuxDistro struct {
+}
+
+func (d *clearLinuxDistro) getID() string {
+	return "clearlinux"
+}
+
+// Correctly split and format the command, using sudo if appropriate, as a
+// common mechanism for the various package install functions.
+func sudoFormatCommand(command string, packages []string) bool {
+	toInstall := strings.Join(packages[0:], " ")
```

Comment:
Note we can likely simplify this and remove the sudo treatment now, as long as euid == 0 is always maintained during component initialization.